src/transformers/models/gptj/modeling_gptj.py

```diff
@@ -0,0 +1,889 @@
+
+            Indices can be obtained using :class:`transformers.GPTJTokenizer`. See
+            :func:`transformers.PreTrainedTokenizer.encode` and :func:`transformers.PreTrainedTokenizer.__call__` for
+            details.
+
+            `What are input IDs? <../glossary.html#input-ids>`__
+        attention_mask (:obj:`torch.FloatTensor` of shape :obj:`({0})`, `optional`):
+            Mask to avoid performing attention on padding token indices. Mask values selected in ``[0, 1]``:
+
+            - 1 for tokens that are **not masked**,
+            - 0 for tokens that are **masked**.
+
+            `What are attention masks? <../glossary.html#attention-mask>`__
+        token_type_ids (:obj:`torch.LongTensor` of shape :obj:`({0})`, `optional`):
+            Segment token indices to indicate first and second portions of the inputs. Indices are selected in ``[0,
+            1]``:
+
+            - 0 corresponds to a `sentence A` token,
+            - 1 corresponds to a `sentence B` token.
+
+            `What are token type IDs? <../glossary.html#token-type-ids>`_
+        position_ids (:obj:`torch.LongTensor` of shape :obj:`{0}`, `optional`):
+            Indices of positions of each input sequence tokens in the position embeddings. Selected in the range ``[0,
+            config.n_positions - 1]``.
+
+            `What are position IDs? <../glossary.html#position-ids>`_
+        head_mask (:obj:`torch.FloatTensor` of shape :obj:`(num_attention_heads,)` or :obj:`(n_layer, num_attention_heads)`, `optional`):
+            Mask to nullify selected heads of the self-attention modules. Mask values selected in ``[0, 1]``:
+
+            - 1 indicates the head is **not masked**,
+            - 0 indicates the head is **masked**.
+
+        inputs_embeds (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length, n_ctx)`, `optional`):
+            Optionally, instead of passing :obj:`input_ids` you can choose to directly pass an embedded representation.
+            This is useful if you want more control over how to convert `input_ids` indices into associated vectors
+            than the model's internal embedding lookup matrix.
+        output_attentions (:obj:`bool`, `optional`):
+            Whether or not to return the attentions tensors of all attention layers. See ``attentions`` under returned
+            tensors for more detail.
+        output_hidden_states (:obj:`bool`, `optional`):
+            Whether or not to return the hidden states of all layers. See ``hidden_states`` under returned tensors for
+            more detail.
+        return_dict (:obj:`bool`, `optional`):
+            Whether or not to return a :class:`~transformers.file_utils.ModelOutput` instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "The bare gptj Model transformer outputting raw hidden-states without any specific head on top.",
+    GPTJ_START_DOCSTRING,
+)
+class GPTJModel(GPTJPreTrainedModel):
```

Comment:
It'd be great if you can add support for `parallelize()` akin to `GPT2Model`.

Given the size of this model checkpoint, that'll likely be necessary for most folks wanting to perform inference since access to A100 GPUs is very limited.

Reply:
(and, of course, for all consuming classes in this script)

Reply:
I fear you misunderstand; you do not need a A100 to use this model. To fine-tune the model requires approximately 24 GB of VRAM.

`parallelize` is an experimental feature and not one I feel comfortable supporting right now. If efficiency is a key concern you shouldn't be using this library anyways, as the `transformer` library's design philosophy is that readability and usability trumps efficiency. To get the most performance out of this model, I recommend using the [original codebase](https://github.com/kingoflolz/mesh-transformer-jax) instead.

Reply:
Sure, but even 32 GB V100 (p3dn.24xlarge) GPUs are expensive + hard to get, so I think at the very least, it should be possible to perform inference on a 16 GB V100 (p3.16xlarge). I'm not sure what sequence lengths + sharding + precision you used for that 24GB VRAM number for fine-tuning, but that's still good to know.

As you say, philosophically usability trumps efficiency in this library, so I think the ability to at least perform inference on cheaper GPUs ought to be considered when it comes to this library. Feature-wise compatibility with other GPT classes in this library also ought to considered.

Ultimately it's up to you - if it turns out that someone intending to **utilize** your PR is budget-constrained, they'll just be forced to end up making those changes locally by copy-pasting stuff from `modeling_gpt2.py`.

Reply:
1. GPT-J 6B easily fits in a 16 GB GPU for inference or 32 GB GPU for tuning (at FP16). CPU inference, while slow, absolutely remains an option and it isn't hard to get a server with 32 GB of memory. And besides, just because GPT-J 6B is the only model that exists today doesn't mean that other models cannot be created in the future at different scales. This is generic model class PR, not a GPT-J 6B PR.
2. Look, we understand that you are concerned with accessibility of large models. [EleutherAI expects that those looking to use our large language models will need to find beefy hardware to run them.](https://www.eleuther.ai/faq) It is an unfortunate result of scaling, but it is something we cannot help with. You can try to do fancy optimizations and swapping weights in and out of memory but that is incompatible with the `transformers` design philosophy and out of scope for this PR.
3. If you want parallelization today, use Mesh Transformer JAX. Parallelization is out of scope for this PR, and will not be considered.

Reply:
Okay now that I am back to work on my laptop browser and not on email, I can see Stella deleted her follow-up comment (but thankfully visible for all to see by clicking the "three-dots button" since I replied via email to that deleted comment). Just reproducing that deleted comment again here for easy accessibility to others reading this PR:

> Again, I fear you misunderstand. You absolutely can perform inference on a
 16 GB V100 GPU. At no point did I say that you needed 32 GB of memory to
 use this model.

Eric, I do think your comment comes off as very hostile and unwelcoming of public suggestions and comments to a PR on a community repo that I do not believe you own (and side-stepping the original point raised).

I'd rather let @patil-suraj, @sgugger or someone else from Hugging Face pitch in on ensuring some semblance of feature-wise compatibility across `modeling_gpt2.py`, `modeling_gptj.py` and `modeling_gpt_neo.py`. Usability trumps efficiency in **this** library as pointed out by Stella, so even purely by that yardstick, it is expected that `parallelize` (which exists in `modeling_gpt2.py`) would be supported in the remaining two aforementioned scripts. If you do not want to do it in this PR, that's fine. Perhaps some one else will do it in a subsequent PR, since this repo is a community repo, not an Eleuther-AI repo.

I'm not going to respond to any further comments here, thanks again for the PR!

Reply:
I do apologize for being quite direct.

Reply:
@g-karthik I am sorry that Eric has been quite rude to you. That said, I largely agree with the points he has raised.

As I explain [here](https://github.com/huggingface/transformers/pull/13022#issuecomment-900005434), I believe you are mistaken about the size of models and the amount of GPU required to use them.

If you would like to have the `parallelize` feature implemented, you are welcome to submit a PR either to [the source of this PR](www.github.com/stellaathena/transformers) or to the `transformers` library after this PR has been merged. As I said before, `parallelize` is an experimental feature that I do not feel comfortable implementing.

Unfortunately maintaining consistent features across models is a nightmare with `transformers` due to the strict no inheritance policy. I would love for this model (as well as `GPT-Neo`) to inherit from `GPT-2`, or even better add functionality to the `GPT-2` class to allow all of these (nearly identical) models to use the same code base. I agree with you and have lobbied for this before, but my understanding is that HF is not interested in permitting these things. This is a HF design question that is far outside the scope of this PR, and I encourage you to take it up with them.

Reply:
For readers of this thread, as I explain [here](https://github.com/huggingface/transformers/pull/13022#issuecomment-900048630), Stella deleted the original claim that "you absolutely can perform inference on a 16 GB V100 GPU" because it was inaccurate, and is now pointing to an updated claim about using BF16.

Usage of FP16/FP32/BF16 are decisions that consumers of this library will make based on their use-case. Addition of `parallelize` would make it possible for consumers to use FP32 during inference. At the moment however, without `parallelize()`, the consumers of this library are forced to use FP16/BF16 during inference if they wanted to use this model on a node with 16 GB / GPU.

Reply:
I agree here with @StellaAthena,

Implementing the `parallelize` functionality is out of scope of this PR. We can look into it in a follow-up PR :-)

Reply:
IDK @patrickvonplaten, feels pretty in-scope to me since authors say upfront they are aiming for config equivalence with GPT-2 for the introduction of their GPT-J model classes, and you'd think that entails implementing APIs such as `parallelize()` akin to the implementations in GPT-2.

Moreover, given that FP16 generation is kinda poor (as you also note elsewhere), having `parallelize()` would make it possible to use FP32 for generation on a node with 8 GPUs (16 GB V100s). Forcing LM logits to FP32 is also a fine option, but `parallelize()` seems like a better option in that respect.

Reply:
@g-karthik if it's so important to you, and you think it's easy to implement, why don't you do so and submit a PR? I would be happy to include your code.